src/sam_file.cpp

```diff
@@ -101,9 +101,10 @@ void SAMFile::write_sam_record_from_hit_cluster(
             (right_flank_start + flank_size) <= seq.seq.size() ? flank_size :
             seq.seq.size() - right_flank_start;
         const std::string right_flank = seq.seq.substr(right_flank_start, right_flank_length);
+        const std::string hit_strand = (first_hit->get_prg_kmer_strand()) ? "256" : "272";
```

Comment:
There is a 256 offset being added if it is the forward strand or the reverse strand. The 256 bit in the SAM file means secondary alignment, so we are either saying: secondary alignment on forward strand (256) or secondary alignment on the reverse strand (272). Any particular reason to add this offset?

Reply:
The most pressing issue is that I don't think `get_prg_kmer_strand()` tells you if the read mapped forwardly or reversely with respect to the PRG. In the following you have some basic test files and command lines for you to debug the reason. It is basically indexing a PRG with a single gene containing a string of `A`s, and querying this exact gene. `pandora` says the query mapped in RC:

`gene.fa`:
```
>gene
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
```
`query.fa` (same file as `gene.fa`):
```
>query
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
```

`read_index.tsv`:
```
query   strand_test/query.fa
```

Command lines:
```
make_prg from_msa -i strand_test/gene.fa -o strand_test/gene
pandora index strand_test/gene.prg.fa
pandora  compare -o strand_test/compare_out strand_test/gene.prg.fa strand_test/read_index.tsv
```

This is the resulting SAM file (flag is 272, mapped in RC):
```
@PG     ID:pandora      PN:pandora      VN:0.9.1
query[0:50]     272     gene    1{[0, 15)}      254     50=     *       0       0       AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA      *       LF:Z:      RF:Z:
```

If we walk a bit through the code, we will find that [in this line, we add a kmer from the PRG to the index](https://github.com/leoisl/pandora/blob/711b7b0c5cb949fcef6235ad824524becea606e6/src/localPRG.cpp#L727-L729). To add a kmer, we first compute the hash of the kmer and its RC, for example (which mirrors the real execution):
```
hash(AAAAAAAAAAAAAAA) = 900
hash(TTTTTTTTTTTTTTT) = 500
```

and then we select the canonical kmer, which is the kmer with smallest hashed value:
```
hash(AAAAAAAAAAAAAAA) = 900
hash(TTTTTTTTTTTTTTT) = 500 --> this will be the canonical kmer
```

That is what `std::min(kh.first, kh.second)` [in this line](https://github.com/leoisl/pandora/blob/711b7b0c5cb949fcef6235ad824524becea606e6/src/localPRG.cpp#L728) does.  In the next line we have `(kh.first <= kh.second)`, where we basically record if the canonical kmer is from the forward or the reverse strand. In this case will be reverse strand, and it will be what will be returned in `get_prg_kmer_strand()`. So, `get_prg_kmer_strand()` basically tells you if we indexed the forward or reverse kmer from the PRG. It does not tell you if the read mapped in forward or reverse way, it is defined before mapping. As an example, if we map reads:
```
>read_1
AAAAAAAAAAAAAAA
>read_2
TTTTTTTTTTTTTTT
```
to this PRG, `get_prg_kmer_strand()` will always return reverse. 

The same indexing procedure happens for the reads. 

Inferring if the read mapped in a forward or reverse way using canonical indexation of minimisers seems trivial at first sight, but it can get confusing. These are some very simplified situation where the PRG / query is either `AAAAA` or `TTTTT` and we try to infer if the read mapped in FWD or RC:


1. Query `AAAAA` maps to PRG `AAAAA` in FWD:
Case 1:
```
PRG kmer: AAAAA
Strand of indexed PRG kmer: F  (forward kmer, AAAAA, indexed)
Read kmer: AAAAA
Strand of indexed read kmer: F
Read mapping: F (this is the trivial case, I think F+F = F)
```
Case 2:
```
PRG kmer: AAAAA
Strand of indexed PRG kmer: R (reverse kmer, TTTTT, indexed)
Read kmer: AAAAA
Strand of indexed read kmer: R (reverse kmer, TTTTT, indexed)
Read mapping: F (R+R = F?)
```

2. Query `TTTTT` maps to PRG `AAAAA` in REV:
Case 1:
```
PRG kmer: AAAAA
Strand of indexed PRG kmer: F  (forward kmer, AAAAA, indexed)
Read kmer: TTTTT
Strand of indexed read kmer: R (reverse kmer, AAAAA, indexed)
Read mapping: R (F+R = R?)
```

Case 2:
```
PRG kmer: AAAAA
Strand of indexed PRG kmer: R  (reverse kmer, TTTTT, indexed)
Read kmer: TTTTT
Strand of indexed read kmer: R (forward kmer, TTTTT, indexed)
Read mapping: R (R+F = R?)
```

So basically, the truth table we have is (I am not sure if I covered all possibilities here, there might be cases I am missing and thus this truth table might be incorrect, to be checked):
```
F+F=F
R+R=F
F+R=R
R+F=R
```
This is exactly what the [`MinimizerHit::same_strands()`](https://github.com/leoisl/pandora/blob/racon_denovo/include/minihit.h#L30-L33) does, which _I think_ is what should be used here and what would give you the information on how the read maps to the PRG. However, **I am not sure** this actually solves this issue, I think this is the type of problem which is much easier to reason looking at several examples. To note also, don't just look at the strand of the first hit, I think you might need to look at the strand of several hits and call what the majority is calling.

If inferring the strand gets complicated, we have a backup plan: `pandora` infers a multi-sample ML sequence and a per-sample ML sequence. Can then use alignment to infer the gene strandness for each sample...